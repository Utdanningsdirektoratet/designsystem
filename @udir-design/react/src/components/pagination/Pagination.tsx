import {
  Pagination,
  type PaginationProps,
  PaginationButton,
  type PaginationButtonProps,
  PaginationItem,
  type PaginationItemProps,
  PaginationList,
  type PaginationListProps,
  usePagination,
  type UsePaginationProps,
} from '@digdir/designsystemet-react';

export {
  Pagination,
  PaginationProps,
  PaginationButton,
  PaginationButtonProps,
  PaginationItem,
  PaginationItemProps,
  PaginationList,
  PaginationListProps,
  usePagination,
  UsePaginationProps,
};
