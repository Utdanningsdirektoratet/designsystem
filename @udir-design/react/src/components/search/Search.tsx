import {
  Search as DigdirSearch,
  SearchButton,
  type SearchButtonProps,
  SearchClear,
  type SearchClearProps,
  SearchInput,
  type SearchInputProps,
  type SearchProps as DigdirSearchProps,
} from '@digdir/designsystemet-react';
import type {
  ComponentRef,
  ForwardRefExoticComponent,
  RefAttributes,
} from 'react';

type SearchProps = Omit<DigdirSearchProps, 'data-color'>;

const Search = DigdirSearch as ForwardRefExoticComponent<
  SearchProps & RefAttributes<ComponentRef<typeof DigdirSearch>>
> &
  Pick<typeof DigdirSearch, 'Clear' | 'Input' | 'Button'>;

// For some reason this fixes "ComponentSubcomponent" -> "Component.Subcomponent" in Storybook code snippets
Search.displayName = 'Search';

export type {
  SearchButtonProps,
  SearchClearProps,
  SearchInputProps,
  SearchProps,
};
export { Search, SearchButton, SearchClear, SearchInput };
