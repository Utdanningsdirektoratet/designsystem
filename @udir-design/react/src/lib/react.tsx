import styles from './react.module.scss';

export const React: React.FC = () => (
  <div className={styles['container']}>
    <h1>Welcome to React!</h1>
  </div>
);

export default React;
