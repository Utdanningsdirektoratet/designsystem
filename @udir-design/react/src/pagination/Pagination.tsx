export {
  Pagination,
  type PaginationProps,
  PaginationButton,
  type PaginationButtonProps,
  PaginationItem,
  type PaginationItemProps,
  PaginationList,
  type PaginationListProps,
  usePagination,
  type UsePaginationProps,
} from '@digdir/designsystemet-react';
