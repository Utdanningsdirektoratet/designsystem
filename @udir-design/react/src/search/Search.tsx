export { Search, type SearchProps } from '@digdir/designsystemet-react';
