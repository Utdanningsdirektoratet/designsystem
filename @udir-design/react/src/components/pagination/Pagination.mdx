import {
  Meta,
  Controls,
  Primary,
  Canvas,
  ArgTypes,
} from '@storybook/addon-docs/blocks';
import * as PaginationStories from './Pagination.stories';

<Meta of={PaginationStories} />

# Pagination

`Pagination` er en liste av knapper som ofte brukes for å navigere mellom sider av informasjon. Komponenten kan brukes på forskjellige måter, se eksempler under [bruk](#bruk).

Vær oppmerksom på:

- Knappene er som standard `button`-elementer.
- Dersom du bruker paginering for å navigere mellom ulike sider i nettleseren, sørg for at pagineringsknappene blir `a`-elementer. Se eksempel [Paginering med lenker](#paginering-med-lenker).

<Primary />
<Controls of={PaginationStories.Preview} />

## Bruk

`Pagination` er en kontrollert komponent. Det vil si at komponentens tilstand om hvilken side som er aktiv styres utenfra. Vi bruker `usePagination` sammen med de enkelte delkomponentene for å implementere funksjonaliteten.
[Les mer om `usePagination`](/docs/utilities-usepagination--docs).

```tsx
import { Pagination } from '@udir-design/react';

const onChange = (event, page: number) => console.log(`Going to page ${page}`);
const [currentPage, setCurrentPage] = useState(4);
const { pages, prevButtonProps, nextButtonProps, hasNext, hasPrevious } =
  usePagination({
    currentPage,
    setCurrentPage,
    onChange,
    totalPages: 10,
    showPages: 7,
  });

<Pagination>
  <Pagination.List>
    <Pagination.Item>
      <Pagination.Button aria-label="Forrige side" {...prevButtonProps}>
        Forrige
      </Pagination.Button>
    </Pagination.Item>
    {pages.map(({ page, itemKey, buttonProps }) => (
      <Pagination.Item key={itemKey}>
        {typeof page === 'number' && (
          <Pagination.Button {...buttonProps} aria-label={`Side ${page}`}>
            {page}
          </Pagination.Button>
        )}
      </Pagination.Item>
    ))}
    <Pagination.Item>
      <Pagination.Button aria-label="Neste side" {...nextButtonProps}>
        Neste
      </Pagination.Button>
    </Pagination.Item>
  </Pagination.List>
</Pagination>;
```

### Paginering med lenker

Alle del-komponenter støtter `asChild` som kan brukes til å endre `Pagination.Button` til lenker (`a`).

<Canvas of={PaginationStories.WithAnchor} />
