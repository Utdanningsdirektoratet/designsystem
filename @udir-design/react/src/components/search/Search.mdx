import { Meta, Canvas, Controls, Primary } from '@storybook/addon-docs/blocks';
import * as SearchStories from './Search.stories';

<Meta of={SearchStories} />

# Search

`Search` er egnet til å implementere søk som lar brukere raskt finne relevant innhold på et nettsted eller i en applikasjon. Komponenten består av et søkefelt, med eller uten en søkeknapp.

**Passer til å**

- søke med fritekst

**Passer ikke til å**

- søke i en liste med alternativer, bruk heller [`Suggestion`](/docs/components-suggestion--docs)
- søke i innhold som er lett å navigere i

## Slik bruker du Search

<Primary />
<Controls />

```tsx
import { Search } from '@udir-design/react';

// live-søk
<Search>
  <Search.Input aria-label="Søk" onChange={(event) => { /* håndter søk */ }} />
  <Search.Clear />
</Search>

// søk ved enter/knappetrykk
<form onSubmit={(event) => {
  event.preventDefault();
  /* håndter søk */
}}>
  <Search>
    <Search.Input aria-label="Søk" />
    <Search.Clear />
    <Search.Button />
  </Search>
<form>
```

## Eksempler

### Varianter

Du velger selv om du vil bruke `Search.Clear` og `Search.Button`. Vi anbefaler at du bruker `Search.Clear` for å gi brukeren mulighet til å fjerne søket.
Dersom du ikke har `Search.Button` vil input feltet få et søkeikon.
Du kan endre `variant` på `Button` for å tilpasse visningen. Eksemplene nedenfor viser varianter som inkluderer ikon, primær- og sekundærknapp.

<Canvas of={SearchStories.Variants} />

### Med Label

Du kan legge til en `Label` komponent som du kobler sammen med `Search.Input`.
I eksempelet har vi brukt `Field` for å få oppkobling mellom `Label` og `Search.Input`.

<Canvas of={SearchStories.WithLabel} />

### Utføre søkehandling ved enter og klikk

Du kan utføre søkehandling ved både <kbd>Enter</kbd> og klikk ved å putte søket i et `<form>`-element. Søkeknappen har `type="submit"` som standard og fungerer derfor automatisk med `<form>`. Eksempelet nedenfor viser hvordan dette kan gjøres.

<Canvas of={SearchStories.Form} />

### Utføre søkehandling "live"

I mange tilfeller er det mer hensiktsmessig å utføre søkehandling mens brukeren skriver i inputfeltet, heller enn å vente på <kbd>Enter</kbd> eller klikk.

<Canvas of={SearchStories.LiveSearch} />

### Kontrollere Search utenfra

`Search.Input` kan kontrolleres utenfra ved å bruke `value` og `onChange`. Dette er nyttig når du vil at `Search` skal oppdateres basert på endringer i andre komponenter eller tilstander.
I dette tilfellet har vi brukt `Search` i et skjema hvor søk blir utført når brukeren trykker <kbd>Enter</kbd>, på "Søk"-knappen, eller en av hurtigsøk-knappene ([`Chip.Button`](/docs/components-chip--docs)).

<Canvas of={SearchStories.Controlled} />

## Retningslinjer

### Unngå placeholder

`Placeholder` forsvinner når brukerne skriver i feltet. Det er derfor bedre å inkludere hint og viktig informasjon i `label` eller den tilhørende beskrivelsen. Kontrastkravene i WCAG gjelder for `placeholder` også, og da blir teksten så mørk at brukerne kan oppfatte `textfield` som utfylt.

### Tilpass bredden på søkefeltet

Bredden på søkefeltet bør tilsvare lengden på søkeordene brukerne vanligvis skriver inn. Størrelsen på feltet er et viktig signal om hva de kan skrive inn. For eksempel bør et søkefelt for personnumre ha en bredde som tilsvarer et personnummer. Hovedsøkefeltet for et nettsted bør ha plass til at brukerne kan se flere ord samtidig.

### Label

I mange tilfeller er det ikke nødvendig med synlig label på `Search`. Feltets utforming, søkeikon og knapp gjør søket tydelig for brukeren.
Legg til `aria-label` eller `aria-labelledby` dersom du ikke kobler `Search.Input` med en `Label`.

## Tilgjengelighet

### Søk-landemerke

Hvis du vil eksponere søket for hjelpemidler som et ARIA søk-landemerke, legger du det i et element med `role="search"`. Velg det ytterste elementet som er tilknyttet søket (f.eks. `<form role="search">`). Dette kan være nyttig om søket er en viktig del av sidens funksjonalitet, som for eksempel hovedsøket for nettsiden eller hvis søkefeltet er starten av en viktig brukeroppgave. Slik blir det mulig å navigere direkte til søket med hjelpemidler.

### Tastaturnavigering

- <kbd>ESC</kbd> - tømmer søkefeltet for en enhetlig opplevelse på tvers av
  nettlesere.
- <kbd>Tab</kbd> - lar brukere bevege seg mellom inputfelt, clear-knapp og
  søkeknapp.
- <kbd>Enter</kbd> - sender inn søket når fokus er i inputfeltet.
