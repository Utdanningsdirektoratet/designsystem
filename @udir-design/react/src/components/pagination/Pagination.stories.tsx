import type { Meta, StoryObj } from '@storybook/react-vite';
import { Pagination } from './Pagination';
import { useState } from 'react';
import { useArgs } from 'storybook/internal/preview-api';
import { expect, userEvent, within } from 'storybook/test';
import {
  usePagination,
  UsePaginationProps,
} from 'src/utilities/hooks/usePagination/usePagination';
import { Search } from '../search/Search';

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  tags: ['beta', 'digdir'],
  parameters: {
    componentOrigin: {
      originator: 'digdir',
      details: 'Vi har fjernet mulighet for fargevalg.',
    },
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Preview: Story = {
  render(args) {
    const [page, setCurrentPage] = useState(4);
    const { pages, nextButtonProps, prevButtonProps } = usePagination({
      currentPage: page,
      totalPages: 10,
      showPages: 7,
      setCurrentPage,
    });

    return (
      <Pagination aria-label="Sidenavigering" {...args}>
        <Pagination.List>
          <Pagination.Item>
            <Pagination.Button aria-label="Forrige side" {...prevButtonProps}>
              Forrige
            </Pagination.Button>
          </Pagination.Item>
          {pages.map(({ page, itemKey, buttonProps }) => (
            <Pagination.Item key={itemKey}>
              {typeof page === 'number' && (
                <Pagination.Button {...buttonProps} aria-label={`Side ${page}`}>
                  {page}
                </Pagination.Button>
              )}
            </Pagination.Item>
          ))}
          <Pagination.Item>
            <Pagination.Button aria-label="Neste side" {...nextButtonProps}>
              Neste
            </Pagination.Button>
          </Pagination.Item>
        </Pagination.List>
      </Pagination>
    );
  },
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    const pagination = canvas.getByRole('navigation');
    const prevButton = canvas.getByRole('button', { name: /forrige side/i });
    const nextButton = canvas.getByRole('button', { name: /neste side/i });

    await step('Pagination component is rendered', async () => {
      expect(pagination).toBeInTheDocument();
    });

    await step('Navigation buttons are rendered', async () => {
      expect(prevButton).toBeInTheDocument();
      expect(nextButton).toBeInTheDocument();
    });

    await step('Initial active page is 4', async () => {
      const activePageButton = canvas.getByRole('button', { name: /side 4/i });
      expect(activePageButton).toHaveAttribute('aria-current', 'page');
    });

    await step('Clicking "Neste side" changes active page to 5', async () => {
      await userEvent.click(nextButton);
      const activePageButton = canvas.getByRole('button', { name: /side 5/i });
      expect(activePageButton).toHaveAttribute('aria-current', 'page');
    });

    await step(
      'Clicking "Forrige side" changes active page back to 4',
      async () => {
        await userEvent.click(prevButton);
        const activePageButton = canvas.getByRole('button', {
          name: /side 4/i,
        });
        expect(activePageButton).toHaveAttribute('aria-current', 'page');
      },
    );
  },
};

type WithAnchorPaginationProps = UsePaginationProps & {
  url: string;
};

export const WithAnchor: StoryObj<WithAnchorPaginationProps> = {
  args: {
    currentPage: 2,
    totalPages: 10,
    showPages: 7,
    url: 'https://udir.design.no/pagination',
  },
  render(args) {
    const [, updateArgs] = useArgs();
    const { pages, nextButtonProps, prevButtonProps } = usePagination({
      ...args,
    });
    const handleAnchorClick = (
      e: React.MouseEvent<HTMLAnchorElement>,
      page: number,
    ) => {
      e.preventDefault();
      if (page === 0 || page > args.totalPages) {
        return;
      }
      const newUrl = `https://udir.design.no/pagination#side-${page}`;
      updateArgs({ currentPage: page, url: newUrl });
      return;
    };

    return (
      <div
        style={{
          gap: 70,
          display: 'flex',
          flexDirection: 'column',
        }}
      >
        <Search>
          <Search.Input aria-label="SÃ¸k" value={args.url} />
        </Search>
        <Pagination aria-label="Sidenavigering">
          <Pagination.List>
            <Pagination.Item>
              <Pagination.Button
                asChild
                aria-label="Forrige side"
                {...prevButtonProps}
              >
                <a
                  onClick={(e) => handleAnchorClick(e, args.currentPage - 1)}
                  href={`#side-${args.currentPage - 1}`}
                >
                  Forrige
                </a>
              </Pagination.Button>
            </Pagination.Item>
            {pages.map(({ page, itemKey, buttonProps }) => (
              <Pagination.Item key={itemKey}>
                {typeof page === 'number' && (
                  <Pagination.Button
                    asChild
                    aria-label={`Side ${page}`}
                    {...buttonProps}
                  >
                    <a
                      href={`#side-${page}`}
                      onClick={(e) => handleAnchorClick(e, page)}
                    >
                      {page}
                    </a>
                  </Pagination.Button>
                )}
              </Pagination.Item>
            ))}
            <Pagination.Item>
              <Pagination.Button
                asChild
                aria-label="Neste side"
                {...nextButtonProps}
              >
                <a
                  onClick={(e) => handleAnchorClick(e, args.currentPage + 1)}
                  href={`#side-${args.currentPage + 1}`}
                >
                  Neste
                </a>
              </Pagination.Button>
            </Pagination.Item>
          </Pagination.List>
        </Pagination>
      </div>
    );
  },
};

export const Mobile: Story = {
  render() {
    return (
      <Pagination>
        <Pagination.List>
          <Pagination.Item>
            <Pagination.Button aria-label="Forrige side" variant="tertiary" />
          </Pagination.Item>

          <Pagination.Item>
            <Pagination.Button aria-label="Side 2" variant="tertiary">
              2
            </Pagination.Button>
          </Pagination.Item>

          <Pagination.Item>
            <Pagination.Button aria-label="Side 3">3</Pagination.Button>
          </Pagination.Item>

          <Pagination.Item>
            <Pagination.Button aria-label="Side 4" variant="tertiary">
              4
            </Pagination.Button>
          </Pagination.Item>

          <Pagination.Item>
            <Pagination.Button aria-label="Neste side" variant="tertiary" />
          </Pagination.Item>
        </Pagination.List>
      </Pagination>
    );
  },
};
