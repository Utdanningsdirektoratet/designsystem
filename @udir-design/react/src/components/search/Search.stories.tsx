import type { Meta, StoryObj } from '@storybook/react';
import { Search } from './Search';

const meta: Meta<typeof Search> = {
  component: Search,
};

export default meta;
type Story = StoryObj<typeof Search>;

export const Preview: Story = {
  render: (args) => (
    <Search {...args}>
      <Search.Input aria-label="SÃ¸k" />
      <Search.Clear />
      <Search.Button />
    </Search>
  ),
};
