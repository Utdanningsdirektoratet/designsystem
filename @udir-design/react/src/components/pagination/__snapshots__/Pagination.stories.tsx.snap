// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mobile 1`] = `
<nav aria-label="Sidenavigering">
  <ul>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Forrige side"
      >
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 2"
      >
        2
      </button>
    </li>
    <li>
      <button
        data-variant="primary"
        type="button"
        aria-label="Side 3"
      >
        3
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 4"
      >
        4
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Neste side"
      >
      </button>
    </li>
  </ul>
</nav>
`;

exports[`Preview 1`] = `
<nav aria-label="Sidenavigering">
  <ul>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Forrige side"
        aria-hidden="false"
      >
        Forrige
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 1"
      >
        1
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 2"
      >
        2
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 3"
      >
        3
      </button>
    </li>
    <li>
      <button
        data-variant="primary"
        type="button"
        aria-label="Side 4"
        aria-current="page"
      >
        4
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 5"
      >
        5
      </button>
    </li>
    <li>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Side 10"
      >
        10
      </button>
    </li>
    <li>
      <button
        data-variant="tertiary"
        type="button"
        aria-label="Neste side"
        aria-hidden="false"
      >
        Neste
      </button>
    </li>
  </ul>
</nav>
`;

exports[`With Anchor 1`] = `
<div style="gap: 70px; display: flex; flex-direction: column;">
  <div>
    <input
      placeholder
      aria-label="SÃ¸k"
      type="search"
      value="https://udir.design.no/pagination"
    >
  </div>
  <nav aria-label="Sidenavigering">
    <ul>
      <li>
        <a
          href="#side-1"
          data-variant="tertiary"
          aria-label="Forrige side"
          aria-hidden="false"
        >
          Forrige
        </a>
      </li>
      <li>
        <a
          href="#side-1"
          data-variant="tertiary"
          aria-label="Side 1"
        >
          1
        </a>
      </li>
      <li>
        <a
          href="#side-2"
          data-variant="primary"
          aria-label="Side 2"
          aria-current="page"
        >
          2
        </a>
      </li>
      <li>
        <a
          href="#side-3"
          data-variant="tertiary"
          aria-label="Side 3"
        >
          3
        </a>
      </li>
      <li>
        <a
          href="#side-4"
          data-variant="tertiary"
          aria-label="Side 4"
        >
          4
        </a>
      </li>
      <li>
        <a
          href="#side-5"
          data-variant="tertiary"
          aria-label="Side 5"
        >
          5
        </a>
      </li>
      <li>
      </li>
      <li>
        <a
          href="#side-10"
          data-variant="tertiary"
          aria-label="Side 10"
        >
          10
        </a>
      </li>
      <li>
        <a
          href="#side-3"
          data-variant="tertiary"
          aria-label="Neste side"
          aria-hidden="false"
        >
          Neste
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
